Sept-Oct Release Notes (#2191)
pulumi/pulumi-hugo: themes/default/content/blog/pulumi-release-notes-80/index.md

  - [New Resources in our Providers](#new-resources-in-our-providers)
  - [Pulumi+Kubernetes Improvements](#pulumikubernetes-improvements)
- Pulumi CLI and core technologies
  - [Pulumi CLI now displays time elapsed per resource](#pulumi-cli-now-displays-time-elapsed-per-resource)
  - [Automation API NodeJS parallel inline programs](#automation-api-nodejs-parallel-inline-programs)
  - [Add Output.format to the Python SDK](#add-outputformat-to-the-python-sdk)
  - [Cloning Pulumi templates from Azure DevOps](#cloning-pulumi-templates-from-azure-devops)
  - [Skip Checkpoints Experimental Flag](#skip-checkpoints-experimental-flag)
  - [Token Authentication in Go Providers](#token-authentication-in-go-providers)
  - [Bump TypeScript version from v3.7.3 to v3.8.3](#bump-typescript-version-from-v373-to-v383)
  - [Display outputs last in diff view](#display-outputs-last-in-diff-view)
  - [Support lazy-loading Node modules](#support-lazy-loading-node-modules)
- Pulumi Service & Pulumi.com
  - [Pulumi Deployments](#pulumi-deployments)
  - [Architecture Templates Support](#architecture-templates-support)

## Cloud Providers and Packages

### New resources in our providers

> We shipped new versions of the AWS Native provider, Google Native provider and the Azure Native provider that added support for 698 new resources in the last two months. 662 of the new resources were added in the Google Native provider, 17 new resources when excluding IAM policy binding improvements, 29 resources were added to the AWS Native provider and 7 were added to the Azure Native provider.

Reviewer: This 698 numbers sounds slightly suspiciously large.  If I'm reading right, that because 645 of them were IAM resources?  Probably fine, but maybe worth commenting on why 645 IAM resources showed up?

Reviewer 2: Yeah, I had the same thought - we can mention IAM resources explicitly and link to the PR.

### Pulumi+Kubernetes Improvements

In October we announced a set of major updates which deepen and extend Pulumi’s support for Kubernetes and the Kubernetes ecosystem. Here are a few of these exciting enhancements:

- [Pulumi Kubernetes Operator v1.10](https://github.com/pulumi/pulumi-kubernetes-operator/#readme): New integration with Flux for richer GitOps support, and ability to deploy Pulumi stacks from directly within the Kubernetes resource model
- [New Pulumi Provider for Flux](https://www.pulumi.com/registry/packages/flux/): Manage Flux with Infrastructure as Code
- [Pulumi Kubernetes Provider v3.22](/registry/packages/kubernetes): Server Side Apply and Resource Patch
  
👉  Learn more in the [Pulumi+Kubernetes: New Flux Integration and Inline Programs blog](/blog/pulumi-kubernetes-new-2022).

> ## Cloud Engineering Days Launches

Reviewer: Should we put this first to match the ToC?  (Or reorder the ToC?)

> In the Pulumi Cloud Engineering Days 2022 we are announced a set of important new advancements in the Pulumi platform designed to help organizations scale with their infrastructure as code needs. These announcements were as follows:

Reviewer 3: ```suggestion
During Pulumi Cloud Engineering Days 2022 we announced a set of important new advancements in the Pulumi platform designed to help organizations scale with their infrastructure as code needs. These announcements were as follows:
```

👉  Learn more in the [Cloud Engineering Days Announcements blog](/blog/nov-2022-launches).

## Pulumi CLI and core technologies

### Pulumi CLI now displays time elapsed per resource

When you modify a resource the Pulumi CLI will display the time elapsed so you can determine how long the modification elapsed for. This change helps surface for each resource the time it takes to create or modify allowing you to plan for it next time. See it in action below!

[![Watch it in action](https://asciinema.org/a/XHYWenULkugoY57kStsadcBH8.svg)](https://asciinema.org/a/XHYWenULkugoY57kStsadcBH8)

👉  Learn more in the [Elapsed Time per Resource blog post](blog/pulumi-cli-time-elapsed) and in the [resource creation times GitHub issue](https://github.com/pulumi/pulumi/issues/10534).

### Automation API NodeJS parallel inline programs

> Automation API now supports parallel execution of NodeJS inline programs in addition to the existing Python inline program support. Now users can run multiple NodeJS Pulumi programs in parallel when they are using Automation API. It uses NodeJS AsyncLocalContext to isolate state away from process environments and into context-specific stores owned by LocalWorkspaces.

Reviewer 2: You can also mention .NET as a supported runtime for parallel execution (it has worked since a long time ago)

👉  Learn more in the [support parallel execution of NodeJS inline programs GitHub issue](https://github.com/pulumi/pulumi/issues/5449).

### New --remove flag for `pulumi destroy`

> A [`pulumi destroy`](/docs/reference/cli/pulumi_destroy) operation destroys all existing resources in the stack, but not the stack itself. If you then wanted to delete the stack itself, once all the resources were destoryed, you would run a [`pulumi stack rm`]. A community member, [mrod-io](https://github.com/mrod-io) added a flag for `pulumi destroy` for when you want to remove the stack after its resources are destroyed: `pulumi destroy --remove`.

Reviewer 3: ```suggestion
A [`pulumi destroy`](/docs/reference/cli/pulumi_destroy) operation destroys all existing resources in the stack, but not the stack itself. If you then wanted to delete the stack itself, once all the resources were destroyed, you would run a [`pulumi stack rm`]. A community member, [mrod-io](https://github.com/mrod-io) added a flag for `pulumi destroy` for when you want to remove the stack after its resources are destroyed: `pulumi destroy --remove`.
```

**See it in action below:**

```console
meagancojocar@Meagans-MacBook-Pro demo % pulumi destroy --remove
Previewing destroy (demo/dev)

View Live: https://app.pulumi.com/meagan/demo/dev/previews/632658ed-41e0-411d-b8a8-f41cc6122ef4

     Type                 Name                  Plan       
 -   pulumi:pulumi:Stack  dev                   delete     
 -   └─ aws:s3:Bucket     my-bucket             delete     
 
Outputs:
  - bucketName: "my-bucket-ad305ba"

Resources:
    - 2 to delete

Do you want to perform this destroy? yes
Destroying (demo/dev)
```

👉  Learn more in the [Add --remove as an option for `pulumi destroy` GitHub issue](https://github.com/pulumi/pulumi/issues/10484).

### Add Output.format to the Python SDK

> We added Output.format to the Python SDK to match helper methods in our other SDKs. Most languages have a standard way to format a string (`String.Format` in C#, `fmt.Sprintf` in Go, etc) but you can't use those functions with our `Output` values, therefore, we have a format method that similar to the languages normal format method, but with support for outputs. We have added `Output.format` in Python which matches the [standard Python `str.format` method](https://docs.python.org/3/library/stdtypes.html#str.format). Similar to how we have `interpolate` in TypeScript, and `Output.Format` in C#.

Reviewer 3: ```suggestion
We added Output.format to the Python SDK to match helper methods in our other SDKs. Most languages have a standard way to format a string (`String.Format` in C#, `fmt.Sprintf` in Go, etc) but you can't use those functions with our `Output` values. We have added a format method that's similar to each language's normal format method, but with support for outputs. We have added `Output.format` in Python which matches the [standard Python `str.format` method](https://docs.python.org/3/library/stdtypes.html#str.format). This is similar to how we have `interpolate` in TypeScript, and `Output.Format` in C#.
```

👉 Learn more by reviewing the [Skip checkpoints pull request](https://github.com/pulumi/pulumi/pull/10750).

### Token Authentication in Go Providers

Thank you to community member [@avlllo](https://github.com/avlllo) for adding support for token authentication in the Go providers that use git. You can now authenticated requests by tokens such as `https://auth:token@gitlab.example.com/group/proj.git/tree/v0.0.0/path`.

👉 Learn more by reviewing the [Add token to git Go module pull request](https://github.com/pulumi/pulumi/pull/10628).

### Bump TypeScript version from v3.7.3 to v3.8.3

> We updated the vendored version of TypeScript in the NodeJS SDK and runtime from v.3.7.3 to v3.8.3. Updating TypeScript versions enables you to use newer TypeScript features, such as the `import type` ans `export type` syntax, which is required to use OpenTelemetry.

Reviewer 3: ```suggestion
We updated the vendored version of TypeScript in the NodeJS SDK and runtime from v.3.7.3 to v3.8.3. Updating TypeScript versions enables you to use newer TypeScript features, such as the `import type` and` export type` syntax, which is required to use OpenTelemetry.
```

👉 Learn more by reviewing the [Bump TSC version to 3.8.3 pull request](https://github.com/pulumi/pulumi/pull/10618).

> ### Display outputs last in diff view

Reviewer: This one is honestly just a bug fix - it really just highlights that we used to do something that makes no sense :-).  I'd be fine leaving it out (especially since it is visually prominent due to the size of the code blocks).  But don't feel strongly - it is indeed a nice fix.

Do you want to perform this update? details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::iac-workshop::pulumi:pulumi:Stack::iac-workshop-dev]
    ~ aws:s3/bucket:Bucket: (update)
        [id=my-bucket-02d7e7a]
        [urn=urn:pulumi:dev::iac-workshop::aws:s3/bucket:Bucket::my-bucket]
        [provider=urn:pulumi:dev::iac-workshop::pulumi:providers:aws::default_1_7_0::229428dd-3bcc-4dcf-ae56-ded0b3b0f322]
      + website: {
          + indexDocument: "index.html"
        }
    ~ aws:s3/bucketObject:BucketObject: (update)
        [id=index.html]
        [urn=urn:pulumi:dev::iac-workshop::aws:s3/bucketObject:BucketObject::index.html]
        [provider=urn:pulumi:dev::iac-workshop::pulumi:providers:aws::default_1_7_0::229428dd-3bcc-4dcf-ae56-ded0b3b0f322]
      ~ acl        : "private" => "public-read"
      ~ contentType: "binary/octet-stream" => "text/html"

    --outputs:--
  + bucketEndpoint: output<string>
  ~ bucketName    : "my-bucket-02d7e7a" => output<string>

Do you want to perform this update? no
```

👉 Learn more by reviewing the [Display outputs last in diff view pull request](https://github.com/pulumi/pulumi/pull/10535)

### Support lazy-loading Node modules

> We now have opt-in support for generating lazy load code for Node modules. Modules that define functions and resources are loaded lazily when you opt-in. Generated code may use `import type ..` form when importing enums ("useTypeOnlyReferences" opt-in flag in schema).

Reviewer: Is this being used anywhere yet?  Might be worth noting either that it is being used in Azure Native, or that we plan to use it in Azure Native and other large Pulumi Packages to improve performance when using these libraries?

4. **git push to Deploy** - Configure automatic infrastructure deployments in response to git push events by installing and configuring the Pulumi GitHub App.
5. **Remote Automation API** - Our strongly typed infrastructure orchestration technology can now run deployments remotely in the cloud.

Click to Deploy
![Click to Refresh in Service](refresh.gif)

Pulumi Deployments removes the cloud as a bottleneck and unlocks it as an accelerant for your business. It is available in preview today. Deployments are free during the preview, with usage-based pricing to come. Check out these resources to get started today:

- [Request access](/product/pulumi-deployments) to the Pulumi Deployments preview
- [The Pulumi Deployments demo video](https://www.youtube.com/watch?v=v48U7CNWutc&ab_channel=PulumiTV)
- [Pulumi Deployments documentation](/docs/intro/pulumi-service/deployments)
- [Pulumi Deployments REST API documentation](/docs/reference/deployments-rest-api)
- [Automation API examples](https://github.com/pulumi/automation-api-examples)

👉  Learn more in the [Pulumi Deployments launch blog post](/blog/pulumi-deployments).

### Architecture Templates Support

> We added Architecture Templates support to the new project experience in the Pulumi Service console. Now in addition to selecting your cloud provider and language of choice you can select a template for your new project. We currently have Static Website, Serverless, Container Service and Virtual Machine templates with more to come!

Reviewer 3: ```suggestion
We added Architecture Templates support to the new project experience in the Pulumi Service console. Now in addition to selecting your cloud provider and language of choice you can select a template for your new project. We currently have Static Website, Serverless, Container Service, Virtual Machine, Kubernetes cluster and web application templates, with more to come!
```
